src/pybind/mgr/dashboard/frontend/src/app/shared/datatable/crud-table/crud-form/array.type.ts

```diff
@@ -0,0 +1,41 @@
+import { Component } from '@angular/core';
+import { FieldArrayType } from '@ngx-formly/core';
+import { forEach } from 'lodash';
+
+@Component({
+  selector: 'cd-formly-array-type',
+  template: `
+    <div class="mb-3">
+      <legend *ngIf="props.label" class="cd-header mt-1">{{ props.label }}</legend>
+      <p *ngIf="props.description">{{ props.description }}</p>
+      <div class="d-flex flex-row-reverse">
+        <button class="btn btn-primary" type="button" (click)="addWrapper()">+</button>
+      </div>
+
+      <div class="alert alert-danger" role="alert" *ngIf="showError && formControl.errors">
+        <formly-validation-message [field]="field"></formly-validation-message>
+      </div>
+
+      <div *ngFor="let field of field.fieldGroup; let i = index" class="row align-items-start">
+        <formly-field class="col" [field]="field"></formly-field>
+        <div *ngIf="field.props.removable !== false" class="col-2 text-right">
+          <button class="btn btn-danger" type="button" (click)="remove(i)">-</button>
+        </div>
+      </div>
+    </div>
+  `
+})
+export class ArrayTypeComponent extends FieldArrayType {
+  addWrapper() {
+    this.add();
+    forEach(this.field.fieldGroup, (field) => {
+      if (field.type == 'object') {
+        field.props.templateOptions = this.props.templateOptions.objectTemplateOptions;
+      }
+    });
```

Comment:
I think this should be separated into an angular component. With html and ts in separate files.